File: spoke-dev.py
#!/usr/bin/env python3
""" This script requests a spoke supernet and updates phpIPAM.

This script requests a spoke supernet based on the AWS region and
updates phpIPAM accordingly, including the creation of further subnets and
reserving IP addresses.

--

This program is free software: you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation, either version 3 of the License, or (at your option) any later
version.

This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
You should have received a copy of the GNU General Public License along with
this program. If not, see <http://www.gnu.org/licenses/>.
"""

__author__ = "Peet van de Sande"
__contact__ = "pvandesande@tmhcc.com"
__license__ = "GPLv3"

import sys
import argparse
import requests
import json
import time
from jinja2 import Template

starttime = time.time()
config = {}
regionalSettings = {
    "ap-southeast-2": {
        "network": 106,
        "dns": 2,
        "tgwId": "tgw-0fc230fd5535b3ddf",
        "tgwInspectionAttachment": "",
        "tgwMainRouteTable": "",
        "tgwInspectionRouteTable": "",
        "dhcpOptions": ""
    },
    "us-east-1": {
        "network": 74,
        "dns": 2,
        "tgwId": "tgw-0031a74e3b340a704",
        "tgwInspectionAttachment": "",
        "tgwMainRouteTable": "",
        "tgwInspectionRouteTable": "",
        "dhcpOptions": ""
    },
    "eu-west-1": {
        "network": 76,
        "dns": 3,
        "tgwId": "tgw-0031a74e3b340a704",
        "tgwInspectionAttachment": "",
        "tgwMainRouteTable": "",
        "tgwInspectionRouteTable": "",
        "dhcpOptions": ""
    },
    "eu-west-2": {
        "network": 75,
        "dns": 3,
        "tgwId": "tgw-000816d04ea49d358",
        "tgwInspectionAttachment": "",
        "tgwMainRouteTable": "",
        "tgwInspectionRouteTable": "",
        "dhcpOptions": ""
    },
    "eu-central-1": {
        "network": 918,
        "dns": 3,
        "tgwId": "tgw-06173001949ff1ea2",
        "tgwInspectionAttachment": "tgw-attach-068d1df133ade8cac",
        "tgwMainRouteTable": "tgw-rtb-05f55e0d134692083",
        "tgwInspectionRouteTable": "tgw-rtb-05f55e0d134692083",
        "dhcpOptions": "dopt-0a11e07c9afdbb7d8"
    }
}

SPOKESIZE = 22 # Supernet size for a standard spoke

def loadConfig():
    '''
    Load a configuration file
    '''

    global config

    try:
        with open('config.json') as config_file:
            config = json.load(config_file)
    except IOError:
        try:
            with open('/etc/netops/phpipam/config.json') as config_file:
                config = json.load(config_file)
        except IOError:
            sys.exit(json.dumps({'code': 501, 'success': 'false', 'data': {'description': "Can't find config file."}}))

def requestSubnet(masterId, size, description, nameserverId = 0, allowRequests = 1, position = 'first'):
    '''
    Request a subnet from a supernet and update its details

    :param int masterId: Subnet ID of the supernet
    :param int size: Size of the requested subnet in bits (1 - 31)
    :param str description: Human-readable description of what this subnet is used for
    :param int nameserverId: ID of nameserver set in phpIPAM
    :param int allowRequests: Define whether IP addresses may be requested from this subnet
    :param str position: Where in the supernet sits the requested subnet
    :return: JSON object with server response
    :rtype: str
    '''

    global config
    url = f"https://{config['server']}/api/{config['appid']}/subnets/{masterId}/{position}_subnet/{size}/"
    headers = {
            'token': config['token'],
            'Content-Type': 'application/json'
    }
    payload = f"""{{
        \"description\": \"{description}\",
        \"pingSubnet\": \"0\",
        \"allowRequests\": \"{allowRequests}\",
        \"nameserverId\": \"{nameserverId}\"
    }}"""

    return requests.request("POST", url, headers=headers, data=payload).text

def createFirstAddress(subnetId, description, isGateway = 0):
    '''
    Create an IP address in a specific subnet

    :param int subnetId: ID of the subnet
    :param str description: Human-readable description of what this subnet is used for
    :param int isGateway: Define whether the device is a gateway (router)
    :return: JSON object with server response
    :rtype: str
    '''

    global config
    url = f"https://{config['server']}/api/{config['appid']}/addresses/first_free/"
    headers = {
            'token': config['token'],
            'Content-Type': 'application/json'
    }
    payload = f"""{{
        \"subnetId\": \"{subnetId}\",
        \"description\": \"{description}\",
        \"is_gateway\": \"{isGateway}\"
    }}"""

    return requests.request("POST", url, headers=headers, data=payload).text

def createSpoke(region, account, size = 22):
    '''
    Calculate and create subnets plus reserved IP addresses

    :param str region: AWS region
    :param str account: Name of the account
    :param int size: CIDR size of the VPC
    :return: JSON object with server response
    :rtype: str
    '''
    global config, regionalSettings 
    output = {'code': 0, 'success': 'false'}
    output['data'] = []

    description = account + ' VpcCidr'
    r = json.loads(requestSubnet(regionalSettings[region]['network'], size, description, regionalSettings[region]['dns']))
    if r['code'] == 201:
        tmp = {'id': r['id'], 'subnet': r['data'], 'description': description}
        output['data'].append(tmp)

        description = account + ' Private subnet AZ A'
        privatea = json.loads(requestSubnet(r['id'], 24, description, regionalSettings[region]['dns']))
        tmp = {'id': privatea['id'], 'subnet': privatea['data'], 'description': description}
        output['data'].append(tmp)
        privatea_gw = json.loads(createFirstAddress(privatea['id'], 'Default gateway', 1))
        privatea_dns = json.loads(createFirstAddress(privatea['id'], 'AWS DNS'))
        privatea_res3 = json.loads(createFirstAddress(privatea['id'], 'Reserved by AWS'))

        description = account + ' Private subnet AZ B'
        privateb = json.loads(requestSubnet(r['id'], 24, description, regionalSettings[region]['dns']))
        tmp = {'id': privateb['id'], 'subnet': privateb['data'], 'description': description}
        output['data'].append(tmp)
        privateb_gw = json.loads(createFirstAddress(privateb['id'], 'Default gateway', 1))
        privateb_res2 = json.loads(createFirstAddress(privateb['id'], 'Reserved by AWS'))
        privateb_res3 = json.loads(createFirstAddress(privateb['id'], 'Reserved by AWS'))

        description = account + ' Transit subnet AZ B'
        transitb = json.loads(requestSubnet(r['id'], 28, description, 0, 0, 'last'))
        tmp = {'id': transitb['id'], 'subnet': transitb['data'], 'description': description}
        output['data'].append(tmp)
        transitb_gw = json.loads(createFirstAddress(transitb['id'], 'Default gateway', 1))
        transitb_res2 = json.loads(createFirstAddress(transitb['id'], 'Reserved by AWS'))
        transitb_res3 = json.loads(createFirstAddress(transitb['id'], 'Reserved by AWS'))

        description = account + ' Transit subnet AZ A'
        transita = json.loads(requestSubnet(r['id'], 28, description, 0, 0, 'last'))
        tmp = {'id': transita['id'], 'subnet': transita['data'], 'description': description}
        output['data'].append(tmp)
        transita_gw = json.loads(createFirstAddress(transita['id'], 'Default gateway', 1))
        transita_res2 = json.loads(createFirstAddress(transita['id'], 'Reserved by AWS'))
        transita_res3 = json.loads(createFirstAddress(transita['id'], 'Reserved by AWS'))
        output['code'] = 200
        output['success'] = 'true'
    else:
        output['code'] = 500
        output['success'] = 'false'
    return output

def renderTemplate(region, account, ipam, template):
    '''
    Render a template to produce the buildspec

    :param str region: AWS region
    :param str account: Account name
    :param str ipam: Dictionary with subnets and related information
    :param str template: Template file
    :return: string
    :rtype: str
    '''

    global config, regionalSettings 
    headers = {
            'token': config['token'],
            'Content-Type': 'application/json'
    }
    payload = ''

    url = f"https://{config['server']}/api/{config['appid']}/tools/nameservers/{regionalSettings[region]['dns']}/"
    r = json.loads(requests.request("GET", url, headers=headers, data=payload).text)
    nameservers = r['data']['namesrv1'].split(';')

    tpl = Template(template)
    return tpl.render (
        nameservers = nameservers,
        account = account,
        region = region,
        vpcCidr = ipam['data'][0]['subnet'],
        privateAIp = ipam['data'][1]['subnet'],
        privateADescription = ipam['data'][1]['description'],
        privateBIp = ipam['data'][2]['subnet'],
        privateBDescription = ipam['data'][2]['description'],
        transitBIp = ipam['data'][3]['subnet'],
        transitBDescription = ipam['data'][3]['description'],
        transitAIp = ipam['data'][4]['subnet'],
        transitADescription = ipam['data'][4]['description'],
        transitGatewayId = regionalSettings[region]['tgwId'],
        transitGatewayInspectionAttachment = regionalSettings[region]['tgwInspectionAttachment'],
        transitGatewayMainRouteTable = regionalSettings[region]['tgwMainRouteTable'],
        transitGatewayInspectionRouteTable = regionalSettings[region]['tgwInspectionRouteTable'],
        dhcpOptionsId = regionalSettings[region]['dhcpOptions']
    )

def main():
    '''
    Main script logic
    '''

    loadConfig()

    global config, regionalSettings

    # Read input arguments
    argp = argparse.ArgumentParser(description = 'Create spoke network, request IP addresses from phpIPAM. Produce buildspec based on input template.')
    argp.add_argument('region', type=str, help='AWS region where the spoke resides')
    argp.add_argument('account', type=str, help='Account name')
    argp.add_argument('template', type=str, help='Template file name')
    args = argp.parse_args()
    region = args.region
    account = args.account
    template = args.template

    output = {'code': 0, 'success': 'false'}
    output['data'] = []

    if region in regionalSettings:
        ipam = createSpoke(region, account, SPOKESIZE)
        if ipam['code'] == 200:
            output['code'] = 200
            output['success'] = 'true'
            output['data'] = ipam['data']
            with open(template) as infile:
                output['buildspec'] = renderTemplate(region, account, ipam, infile.read())
        else:
            output['code'] = 500
            output['success'] = 'false'
    else:
        output['data'] = {'description': 'Region not defined or recognised.'}

    output['time'] = time.time() - starttime
    return output

if __name__ == "__main__":
    print(json.dumps(main()))

File: spoke-dev.php
<?php
$json = json_decode(file_get_contents("php://input"));

// if (count($json) > 0) {
if (isset($json)) {
  print_r( buildspec($json->region, $json->account));
} elseif (isset($_POST['region']) && (strlen($_POST['account']) > 1)) {
  $buildspec = json_decode(buildspec($_POST['region'], $_POST['account'], $_POST['format']), true);
  if ($buildspec['success'] == 'true') {
    ob_start('ob_gzhandler');
    switch ($_POST['format']) {
      case 'cf':
        header('Content-type: text/yaml');
        header('Content-disposition: attachment; filename="' . $_POST['account'] . '.yaml"');
        break;
      case 'tf':
        header('Content-type: text/json');
        header('Content-disposition: attachment; filename="' . $_POST['account'] . '.tf"');
        break;
    }
    echo $buildspec['buildspec'];
  } else {
    print_help($buildspec['code'], $buildspec['data']['description']);
  }
} else {
  print_help();
}

function buildspec($region, $account, $format) {
  $output = null;
  $retval = null;
  switch ($format) {
    case 'cf':
      $template = '/var/netops/aws/spoke-dev.yaml';
      break;
    case 'tf':
      $template = '/var/netops/aws/spoke-dev.tf';
      break;
  }
  $cmd = sprintf('/usr/local/bin/spoke-dev.py %s %s %s', $region, $account, $template);
  exec($cmd, $output, $retval);
  if ($retval <> 0) {
    return sprintf("Error %d", $retval);
  } else {
    return $output[0];
  }
}

function print_form() {
  $output = "
  <p>Or use this form to download the buildspec manually.</p>
  <form method=\"post\" action=\"" . $_SERVER['PHP_SELF'] . "\">
    <table>
    <tr>
      <td colspan=2>Select region:<br />
        <input type=\"radio\" name=\"region\" value=\"ap-southeast-2\">Asia Pacific (Sydney) ap-southeast-2<br />
        <input type=\"radio\" name=\"region\" value=\"eu-central-1\">Europe (Frankfurt) eu-central-1<br />
        <input type=\"radio\" name=\"region\" value=\"eu-west-1\">Europe (Ireland) eu-west-1<br />
        <input type=\"radio\" name=\"region\" value=\"eu-west-2\">Europe (London) eu-west-2<br />
        <input type=\"radio\" name=\"region\" value=\"us-east-1\">US East (N. Virginia) us-east-1
      </td>
    </tr>
    <tr>
      <td colspan=2>
        <input type=\"radio\" name=\"format\" value=\"cf\">CloudFormation<br />
        <input type=\"radio\" name=\"format\" value=\"tf\" checked=\"checked\">Terraform
      </td>
    </tr>
    <tr>
      <td>Account Name:</td>
      <td><input type=\"text\" name=\"accountName\"></td>
    </tr>
    <tr>
      <td>Account Number:</td>
      <td><input type=\"text\" name=\"accountNumber\"></td>
    </tr>
    <tr>
      <td>Business Segment</td>
      <td><input type=\"text\" name=\"businessSegment\"></td>
    </tr>
    <tr>
      <td>Business Department</td>
      <td><input type=\"text\" name=\"businessDepartment\"></td>
    </tr>
    <tr>
      <td>Cost Center</td>
      <td><input type=\"text\" name=\"costCenter\"></td>
    </tr>
    </table>
    <input type=\"submit\" value=\"Submit\">
  </form>
";
  return $output;
}

function print_help($errcode = 0, $errmsg = '') {
  $output  = "<html>
<head>
  <title>IPAM API</title>
  <script src=\"https://cdn.jsdelivr.net/gh/google/code-prettify@master/loader/run_prettify.js\"></script>
  <style>
    dt {font-weight: bold;}
  </style>
</head>
<body>";
  if ($errmsg !== '') {
    $output .= sprintf("
  <h1>Error %s: %s</h1>", $errcode, $errmsg);
  }
  $output .= " 
  <h1><center>spoke.php</center></h1>
  <br>
  <p>This API requests a subnet for a Spoke VPC and updates <a href=\"https://phpipam.hcch.com/\">phpIPAM</a>.</p>
  <p>All input is expected by POST method encoded as JSON.</p>
  <p>Output is a JSON string containing:
    <ol>
    <li>Response codes, processing time</li>
    <li>ID, IP range and description of VPC</li>
    <li>ID, IP range and description of Private subnet A</li>
    <li>ID, IP range and description of Private subnet B</li>
    <li>ID, IP range and description of Transit subnet B</li>
    <li>ID, IP range and description of Transit subnet A</li>
    <li>YAML to feed into CloudFormation</li>
    </ol>
  </p>
  <p>
  The CloudFormation YAML does the following:<br>
  <ol>
  <li>Creation of the VPC</li>
  <li>Creation of the DHCP Options</li>
  <li>Creation of 2 Private Subnets</li>
  <li>Creation of 2 Transit Subnets</li>
  <li>Creation of 2 Private Route Tables (AZ A, AZ B)</li>
  <li>Creation of the Transit Gateway Attachment</li>
  <li>Configuration of default routes into both Route Tables</li>
  </ol>
  </p>
  <p>The following variables are required:</p>
  <dl>
    <dt>account</dt>
    <dd>The name of the account</dd>
    <dt>region</dt>
    <dd>The region where the spoke will reside</dd>
  </dl>
  <p>Example Python 3 code:</p>
  <pre class=\"prettyprint lang-py\">
import requests

url = \"http://gblon-l-pvd01.hcch.com/api/spoke.php\"
headers = {
  'Authorization': 'Basic bmV0b3BzOkN5dDBza2UhZXRvbg==',
  'Content-Type': 'application/json'
}
payload=\"{\\\"region\\\": \\\"eu-west-1\\\", \\\"account\\\": \\\"Testing123\\\"}\"

response = requests.request(\"POST\", url, headers=headers, data=payload)

print(response.text)
  </pre>
  <br />" . print_form() . "
</body>
</html>";
  print $output;
  exit();
}


?>
